File: algorithms/java/乘积为正数的最长子数组长度/乘积为正数的最长子数组长度.java
// Source : https://leetcode.cn/problems/maximum-length-of-subarray-with-positive-product/
// Author : Yanan Wang
// Date   : 2023-06-23

class Solution {
    public int getMaxLen(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n + 1][2];
        int result = 0;
        for (int i = 1; i <= n; i++) {
            if (nums[i - 1] > 0) {
                dp[i][0] = dp[i - 1][0] + 1;
                dp[i][1] = dp[i - 1][1] == 0 ? 0 : dp[i - 1][1] + 1;
            } else if (nums[i - 1] < 0) {
                dp[i][0] = dp[i - 1][1] == 0 ? 0 : dp[i - 1][1] + 1;
                dp[i][1] = dp[i - 1][0] + 1;
            }  
            result = Math.max(result, dp[i][0]);
        }
        return result;
    }
}

File: algorithms/python/乘积为正数的最长子数组长度/乘积为正数的最长子数组长度.py
# Source : https://leetcode.cn/problems/maximum-length-of-subarray-with-positive-product/description/
# Author : Yanan Wang
# Date   : 2024-01-16

class Solution:
    def getMaxLen(self, nums: List[int]) -> int:
        n = len(nums)
        f = [0] * (n + 1)
        g = [0] * (n + 1)
        ans = 0
 
        for i in range(1, n + 1):
            if nums[i - 1] > 0:
                f[i] = f[i - 1] + 1
                g[i] = g[i - 1] + 1 if g[i - 1] != 0 else 0
            elif nums[i - 1] < 0:
                g[i] = f[i - 1] + 1
                f[i] = g[i - 1] + 1 if g[i - 1] != 0 else 0
            ans = max(ans, f[i])
 
        return ans
